### ISSUE TITLE
`DynamicScene` - Malformed enum when reserializing

### ISSUE THREAD

**Issue by mrchantey** [bevyengine/bevy#12462]
## Bevy version

`0.14.0-dev` `0.13.1` `0.12.1` `0.10.0`

## Relevant system information

- Confirmed on windows and ubuntu wsl

## What you did

1. Serialize a scene containing an enum that has a tuple variant
2. Deserialize it
3. Serialize it again

## What went wrong

Enums use the name of the first variant with the value of another:
```rust
    #[derive(Reflect, Default)]
    enum MyEnum {
        #[default]
        Unit,
        Tuple(String),
        Struct {
            value: u32,
        },
    }
```

Will serialize `MyEnum::Tuple("Hello World")` as `Unit("Hello World")`.

## Steps to reproduce

Add the following to the `should_roundtrip_bincode` test [crates\bevy_scene\src\serde.rs:868](
https://github.com/bevyengine/bevy/blob/ee0fa7d1c22adf5bf4cbac2553e1055b7f3cc579/crates/bevy_scene/src/serde.rs#L868)

```rust
        let scene_serializer2 = SceneSerializer::new(&deserialized_scene, &registry.0);
        let serialized_scene2 = bincode::serialize(&scene_serializer2).unwrap();

        println!(
            "malformed enum: {}",
            deserialized_scene.serialize_ron(&registry.0).unwrap()
        );
        assert_eq!(serialized_scene, serialized_scene2);
```

Will output:

```rust
malformed enum: (
  resources: {},
  entities: {
    4294967296: (
      components: {
        "bevy_scene::serde::tests::MyComponent": (
          foo: (1, 2, 3),
          bar: (1.3, 3.7),
          baz: Unit("Hello World!"),
        ),
      },
    ),
  },
)
```

## Additional information

- Demonstrated using `bincode` but same behavior for `ron`, `serde_json` etc
- I feel like im missing something obvious because it goes at least to version `0.10`, is this intended behavior?

**Comment by alice-i-cecile** [comment]
This seems like a pretty straightforward reflection serde bug. If you submit a patch with a test I'll happily review it.

**Comment by MrGVSV** [comment]
Just tested it and can confirm this is a bug.

<details>
<summary>Full repro</summary>

```rust
#[derive(Reflect, Default)]
enum MyEnum {
    #[default]
    Unit,
    Tuple(String),
    Struct {
        value: u32,
    },
}

let mut registry = TypeRegistry::default();
registry.register::<MyEnum>();

let value = MyEnum::Tuple("Hello world".to_string());

let serializer = ReflectSerializer::new(&value, &registry);
let serialized = ron::ser::to_string(&serializer).unwrap();
println!("{}", serialized);

let mut deserializer = ron::de::Deserializer::from_str(&serialized).unwrap();
let reflect_deserializer = UntypedReflectDeserializer::new(&registry);
let value = reflect_deserializer.deserialize(&mut deserializer).unwrap();
println!("{:?}", value);

let serializer = ReflectSerializer::new(&*value, &registry);
let serialized = ron::ser::to_string(&serializer).unwrap();
println!("{}", serialized);
```

</details>

I think this is happening because we don't pass the type information into the deserialized dynamic output (in this case `DynamicEnum`).

**Comment by MrGVSV** [comment]
Closed by #12464
